src/dune_engine/rule.ml

```diff
@@ -103,13 +98,20 @@ let make ?(mode = Mode.Standard) ~context ?(info = Info.Internal) ~targets actio
            (Dpath.describe_target path))
   in
   let loc =
-    match info with
-    | From_dune_file loc -> loc
-    | Internal ->
-      Loc.in_file
-        (Path.drop_optional_build_context
-           (Path.build (Path.Build.relative dir "_unknown_")))
-    | Source_file_copy p -> Loc.in_file (Path.source p)
+    match loc with
```

Comment:
It feels a little odd to me to have this logic on the inside of `Rule.make`. It makes the logic somewhat hidden/magic from the point of view of callers.

One alternative would be to make the `loc` argument required, and provide functions `internal_rule_loc : dir:Path.t -> Loc.t` and `copy_rule_loc : Path.Source.t -> Loc.t` that factor this hidden functionality out so that it can be used explicitly when needed.

Not a very strong opinion but the current API seems odd to me.